alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/processor/service/CycleService.java

```diff
@@ -0,0 +1,127 @@
+            return null;
+        },false, true);
+
+        if(reachedMaxTx.get()) {
+            try {
+                logger.error("Max transaction reached, entering idle state");
+                Thread.sleep(configuration.getTimeIncrementSeconds() * 1000);
+            } catch (InterruptedException e) {
+                logger.error("Idling has failed, aborting...", e);
+                continueCycle.set(false);
+            }
+        }
+    }
+
+    void start(Cycle cycle) {
+        long txnBatchSize = cycle.getTxnBatchSize();
+        long timeIncrementSec = cycle.getTimeIncrementSeconds();
+
+        logger.debug("Tracking changes ... Start commit time: {}",
+                DateTimeUtil.toReadableString(cycle.getFirstCommitTime()));
+
+        processTxnRange(cycle, txnBatchSize, timeIncrementSec);
+
+        while(txnHistoryIsCatchingUp(timeIncrementSec,cycle.getCurrentCommitTimeMs())){
+            cycle.setCurrentCommitTimeMs(cycle.getCurrentCommitTimeMs() + timeIncrementSec);
+            processTxnRange(cycle, txnBatchSize, timeIncrementSec);
+        }
+    }
+
+    void processTxnRange(Cycle cycle, long txnBatchSize, long timeIncrementSeconds) {
+        // Save current progress in case of transactions collection failure
+        long maxTxId = cycle.getCurrentTransactionId();
+        long maxCommitTimeMs = cycle.getCurrentCommitTimeMs();
+
+        try {
+            List<Transaction> txs = getNodeTransactions(
+                    txnBatchSize, timeIncrementSeconds);
+
+            logger.debug("Found {} transactions", txs.size());
+
+            if(txs.size() > 0) {
+                // Yay, we have a list of transactions to process! Save them for later
+                maxCommitTimeMs = txs.stream().map(Transaction::getCommitTimeMs)
+                        .max(Long::compare).get();
+                maxTxId = txs.stream().map(Transaction::getId)
+                        .max(Long::compare).get();
+
+                for(Transaction tx : txs) {
+                    try {
+                        retryingTransactionHelper.doInTransaction(() -> {
+                            // do work
+                            return null;
+                        }, true, false);
+                    } catch (Exception e) {
+                        logger.error("Tracker loop failed: " + e.getMessage(), e);
+                    }
+                }
+            }
+
+            cycle.setCurrentCommitTimeMs(maxTxId);
+            cycle.setCurrentCommitTimeMs( Long.max(maxCommitTimeMs, cycle.getCurrentCommitTimeMs()));
+
+        } catch (Exception ex) {
+            logger.error("Impossible to read tracker info: " + ex.getMessage(), ex);
+        }
+    }
+
+    boolean txnHistoryIsCatchingUp(long timeIncrementSec, long commitTimeMs) {
+        return DateTimeUtil.xSecondsAgoToMs(timeIncrementSec) > commitTimeMs;
+    }
+
+    private boolean reachedLastTx(long currentTransactionId) {
```

Comment:
How about, along with the current txn id, also bookkeeping the max txn id in the cycle?
Then the cycle is the information expert that can carry this responsibility

Reply:
@driesdeproost this comes in a later branch to fetch the MAX transaction ID from Solr.
Do you still want to add things that will only be handy in between a few commits?

Reply:
yes